### DISCUSSION TITLE
How to import types for react-select v.4?

### DISCUSSION THREAD

**Discussion by lucypoly** [JedWatson/react-select#4927] [Q&A]
Hello I am using `"react-select": "4.3.1",`
But when I try to `import { Props } from 'react-select';`
I receive an error
```TS7016: Could not find a declaration file for module 'react-select'. '/Users/lpolianychko/Projects/lib-components/node_modules/react-select/dist/react-select.cjs.js' implicitly has an 'any' type.   Try `npm install @types/react-select` if it exists or add a new declaration (.d.ts) file containing `declare module 'react-select';```
When I go to **@types/react-select** it is deprecated! How should I import types?

**Comment by Rall3n** [comment]
Hello @lucypoly,

yes, `@types/react-select` has been deprecated, but only for versions `5.x.x` and greater, as they ship with their own types. 

For `v4` it is still recommended to use `@types/react-select`.

PS: Use the version of `@types/react-select` before it's deprecation (`4.0.18`).

**Reply by lucypoly** [reply]
@Rall3n  ok, I tried to use it and import @types in my components library for one of my components. But when I try to use this component in my final app it doesn't see this type, and shows me
```Type '{ className: string; value: { value: string; label: string; id: string; }[]; labelText: string; onChange: (values: unknown) => Promise<void>; }' is not assignable to type 'IntrinsicAttributes & SelectCreatableProps & { children?: ReactNode; }'.   Property 'className' does not exist on type 'IntrinsicAttributes & SelectCreatableProps & { children?: ReactNode; }'.```

In my library it looks like
```
import { Props } from 'react-select/creatable';
export interface ChipProps {
  id?: string;
  value: string;
  label: string;
  disabled?: boolean;
}

export interface SelectCreatableProps extends Props<ChipProps, true> {
  creatable?: boolean;
  disabled?: boolean;
  required?: boolean;
  labelClassName?: string;
  labelText?: ReactNode;
  labelRightContent?: ReactNode;
  onChange?: (args: unknown) => void;
}
```

**Reply by Methuselah96** [reply]
Can you show your usage? It looks like you're providing a `className` as a prop which it's saying isn't a valid prop.

**Reply by lucypoly** [reply]
@Methuselah96  No, for prop `value` I have the same problem, it can't see imported from `react-select` types. For now, I deal with it by adding @types/react-select to my final app as a dev dependency.

**Reply by Methuselah96** [reply]
Yeah, `className` should be a valid prop anyway. It sounds like you're using `react-select` v4, so adding `@types/react-select` as a dev dependency is the correct thing to do.

**Reply by lucypoly** [reply]
@Methuselah96 yes 4, I mean adding dependency not to my lib, where I use react-select, but for final app, that is using my lib

**Reply by Methuselah96** [reply]
Okay, do you have a regular dependency on `@types/react-select` in your lib? That should be the correct way to do it.

**Reply by lucypoly** [reply]
@Methuselah96, I don't think it should be there, I have it in dev dep. Because my lib is dependency itself, so all its dependencies go to dependency in my final app, and that is not ok that we have types in dependency in production. I think should be the way to export them somehow so my app will see it.

**Reply by Methuselah96** [reply]
This is not something specific to `react-select`, libraries need to have a dependency on types package if they use them in their own types. See https://github.com/Microsoft/types-publisher/issues/81 and https://stackoverflow.com/a/46011417/5327429.

It shouldn't be a problem that the type are "in production." If you're bundling, then the types will be not be included in the bundle. If you're not bundling, then types won't make a sizeable difference, they're just extra files that aren't important.
